compat: implement strptime(3) for Windows

fujimotos opened the pull request:
This is a series of patches to make strptime(3) usable on Windows.

bcc55f04 compat: implement strptime(3) for Windows
9815a495 compat: implement langinfo.h for Windows

Note that the implementation is based on the one of musl libc
(I tweaked it so that it works on Windows without problems).
The license is MIT/X, so it should be compatible with fluent-bit.

fujimotos commented:
Part of #960.

bluebike commented:
Actually macOS implmentation of strptime(3) doesn't seem to be full compatible with Linux (glibc) version.  
Actually it would be nice to allow generic use of this private implementation of strptime(3).
There could build option FLB_STRPTIME/FLB_OWN_STRPTIME, etc... 
which could be enabled default in  Win/macOS (?)

edsiper commented:
@fujimotos what do you think about @bluebike comments? (re: macOS)

fujimotos commented:
@edsiper @bluebike 

OK. I'll rework this patch in @bluebike's direction.

> Actually it would be nice to allow generic use of this private implementation of strptime(3).

On that point, I'm a bit doubting that the current version is very useful,
since it (which is based on musl libc) is not fully compatible with glibc
either.

I investigate this issue today to see if I can come up with a more
reasonable solution.

edsiper commented:
@fujimotos we can merge this open and create a new ticket since it's a macOS specific issue (out of the scope of #960

fujimotos commented:
@edsiper Thanks, but no need to merge this PR.

After thinking about this issue in the morning, I'm inclined to try a
different cut:

 1. Use OpenBSD's strptime (for better compatibility with glibc).
 2. Put the implementation into `flb_strptime.h` (not `win32/strptime.h`).

I have a working patch now, and it seems to be working better than the
previous version -- for example, it can handle glibc-specific "%z"
format string without problems.

I will post the new version after more testing; so please wait a bit.

fujimotos commented:
@edsiper @bluebike I've reworked on this patch, and submitted a
new version in the following patches (split into 4 parts).

 - #1105 "langinfo: implement langinfo.h for Windows"
 - #1106 "compat: add a set of macros for implementing strptime"
 - #1109 "strptime: add a new portable strptime(3) (version 2)"
 - #1110 "strptime: set up the build settings for flb_strptime"

I have attached my testing result of this version in https://github.com/fluent/fluent-bit/pull/1109#issuecomment-463491115.
As far as I have tested, it was reasonably compatible with glibc.

Also I added a new compile option FLB_SYSTEM_STRPTIME in #1110.
You can switch to the implementation on platforms other than
Windows by running cmake like below:

    $ cmake .. -DFLB_SYSTEM_STRPTIME=Off
